packages/core/src/core/rename.ts

```diff
@@ -0,0 +1,234 @@
+): Promise<DetailedChange[]> => {
+  const getReferences = (element: Element): { path: ElemID; value: ReferenceExpression }[] => {
+    const references: { path: ElemID; value: ReferenceExpression }[] = []
+    const func: WalkOnFunc = ({ path, value }) => {
+      if (isReferenceExpression(value)
+      && (sourceElemId.isEqual(value.elemID) || sourceElemId.isParentOf(value.elemID))) {
+        references.push({ path, value })
+        return WALK_NEXT_STEP.SKIP
+      }
+      return WALK_NEXT_STEP.RECURSE
+    }
+    walkOnElement({ element, func })
+    return references
+  }
+
+  const references = await awu(await elementsSource.getAll())
+    .flatMap(element => getReferences(element)).toArray()
+
+  return references.map(reference => {
+    const targetReference = new ReferenceExpression(
+      new ElemID(
+        sourceElemId.adapter,
+        sourceElemId.typeName,
+        sourceElemId.idType,
+        targetElemId.name,
+        ...reference.value.elemID.createTopLevelParentID().path
+      ),
+      reference.value.value,
+      reference.value.topLevelParent
+    )
+    return {
+      id: reference.path,
+      action: 'modify',
+      data: {
+        before: reference.value,
+        after: targetReference,
+      },
+    }
+  })
+}
+
+const renameElementPathIndex = async (
+  index: PathIndex,
+  elementFragments: Element[],
+  sourceElemId: ElemID,
+  targetElemId: ElemID
+): Promise<void> => {
+  // The renamed element will be located according to the element's path and not the actual
+  // locations in the nacl. Such that if the user renamed the file before she renamed the Element,
+  // the renamed element will be placed in the original file name. This is because we use and update
+  // the pathIndex and not the ChangeLocation (SourceMap) logic in the current implementation.
+  const pathHints = getElementsPathHints(elementFragments)
+
+  await Promise.all(pathHints.map(entry => index.delete(entry.key)))
+  await Promise.all(pathHints.map(entry => {
+    const elemId = new ElemID(
+      sourceElemId.adapter,
+      sourceElemId.typeName,
+      sourceElemId.idType,
+      targetElemId.name,
+      // this implementation works on InstanceElement only
+      // it won't work on Field elements because they aren't top-level elements
+      ...ElemID.fromFullName(entry.key).createTopLevelParentID().path
+    )
+    return index.set(elemId.getFullName(), entry.value)
+  }))
+}
+
+export const renameElement = async (
+  elementsSource: ElementsSource,
+  sourceElemId: ElemID,
+  targetElemId: ElemID,
+  index?: PathIndex
+): Promise<DetailedChange[]> => {
+  const source = await elementsSource.get(sourceElemId)
+  const elements = isDefined(index)
+    ? await splitElementByPath(source, index)
+    : [source]
+
+  const elementChanges = getRenameElementChanges(sourceElemId, targetElemId, elements)
+  const referencesChanges = await getRenameReferencesChanges(elementsSource, sourceElemId,
+    targetElemId)
```

Comment:
code style nit - don't split lines at a random point
either everything is on the same line, each argument is on a new line, or, less common but I guess also possible, all arguments on the same line
 ```suggestion
  const referencesChanges = await getRenameReferencesChanges(
    elementsSource,
    sourceElemId,
    targetElemId,
  )
```